xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+        object.__setattr__(self, "regions", IList(regions))
+
+        self.operands.freeze()
+        self.results.freeze()
+        self.successors.freeze()
+        self.regions.freeze()
+
+    @classmethod
+    def get(cls, name: str, op_type: type[Operation],
+            operands: Sequence[ISSAValue], result_types: Sequence[Attribute],
+            attributes: dict[str, Attribute], successors: Sequence[IBlock],
+            regions: Sequence[IRegion]) -> IOp:
+        return cls(OpData(name, op_type, attributes), operands, result_types,
+                   successors, regions)
+
+    def __hash__(self) -> int:
+        return hash(id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    @property
+    def name(self) -> str:
+        return self._op_data.name
+
+    @property
+    def op_type(self) -> Type[Operation]:
+        return self._op_data.op_type
+
+    @property
+    def attributes(self) -> dict[str, Attribute]:
+        return self._op_data.attributes
+
+    @property
+    def result(self) -> IResult | None:
+        if len(self.results) > 0:
+            return self.results[0]
+        return None
+
+    @property
+    def region(self) -> IRegion | None:
+        if len(self.regions) > 0:
+            return self.regions[0]
+        return None
+
+    @property
+    def result_types(self) -> List[Attribute]:
+        return [result.typ for result in self.results]
+
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[dict[IBlock, Block]] = None) -> Operation:
+        """
+        Returns a mutable operation that is a copy of this immutable operation.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+        mutable_operands: List[SSAValue] = []
+        for operand in self.operands:
+            if operand in value_mapping:
+                mutable_operands.append(value_mapping[operand])
+            else:
+                print(f"ERROR: op {self.name} uses SSAValue before definition")
+                # Continuing to enable printing the IR including missing
+                # operands for investigation
+                mutable_operands.append(
+                    OpResult(
+                        operand.typ,
+                        None,  # type: ignore
+                        0))
+
+        mutable_successors: List[Block] = []
+        for successor in self.successors:
+            if successor in block_mapping:
+                mutable_successors.append(block_mapping[successor])
+            else:
+                raise Exception("Block used before definition")
```

Comment:
```suggestion
                raise Exception("Invalid IR: Block is not defined in the current region")
```

Reply:
We should maybe create an `InvalidIRException` or something.

Reply:
Yes, I agree. Maybe something like `ScopingException`

Reply:
I think `InvalidIREception` sounds quite a bit more general, so i would personally go with that...

Reply:
I think we should have a hierarchy of them ;)